### PR TITLE
[SPARK-43050][SQL] Fix construct aggregate expressions by replacing grouping functions

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR fixes construct aggregate expressions by replacing grouping functions if a expression is part of aggregation.
In the following example, the second `b` should also be replaced:
image

### Why are the changes needed?

Fix bug:
```
spark-sql (default)> SELECT CASE WHEN a IS NULL THEN count(b) WHEN b IS NULL THEN count(c) END
                   > FROM grouping
                   > GROUP BY GROUPING SETS (a, b, c);
[MISSING_AGGREGATION] The non-aggregating expression "b" is based on columns which are not participating in the GROUP BY clause.
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -610,31 +610,21 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
...
-        case e if isPartOfAggregation(e) => e
+        gid: Attribute): Seq[NamedExpression] = {
+      def replaceExprs(e: Expression): Expression = e match {
+        case e if AggregateExpression.isAggregate(e) => e
         case e =>
           // Replace expression by expand output attribute.
           val index = groupByAliases.indexWhere(_.child.semanticEquals(e))
           if (index == -1) {
-            e
+            e.mapChildren(replaceExprs)
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:620]: Replace `transformDown` with `mapChildren`.

**Comment by wangyum** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -610,31 +610,21 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
...
         groupingAttrs: Seq[Expression],
-        gid: Attribute): Seq[NamedExpression] = aggregations.map { agg =>
-      // collect all the found AggregateExpression, so we can check an expression is part of
-      // any AggregateExpression or not.
-      val aggsBuffer = ArrayBuffer[Expression]()
-      // Returns whether the expression belongs to any expressions in `aggsBuffer` or not.
-      def isPartOfAggregation(e: Expression): Boolean = {
-        aggsBuffer.exists(a => a.exists(_ eq e))
-      }
-      replaceGroupingFunc(agg, groupByExprs, gid).transformDown {
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:614]: why is `transformDown` not working?

[the remark this answers is not in the document]
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:614]: `transformDown` will also transforms aggregate expressions. For expression `b`, the first should not be replaced, the second needs to be replaced. `transformDown` can't do it:
image

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — COMMENTED** [review]
Is this only for `master` branch, @wangyum ?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by wangyum** [comment]
@dongjoon-hyun Both branch-2.x and branch-3.x have this issue.

**Comment by cloud-fan** [comment]
or shall we include this fix in 3.2 as it's the last chance? I don't strong opinion. @dongjoon-hyun can you make a call here since you are the release manager of 3.2.4?

**Comment by dongjoon-hyun** [comment]
I'd prefer to consider this for `branch-3.3`+ because this was not a regression at 3.2, @cloud-fan ~

**Comment by dongjoon-hyun** [comment]
BTW, are you going to block Apache Spark 3.4 RC7, @cloud-fan ?

cc @xinrong-meng , too

**Comment by cloud-fan** [comment]
Since it's not a regression, we don't need to block 3.4 either.

**Comment by dongjoon-hyun** [comment]
Ya, I think so too~

**Comment by dongjoon-hyun** [comment]
This patch can wait for Apache Spark 3.4.1 and 3.3.3.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -603,31 +603,21 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
...
-            e
+            e.mapChildren(replaceExprs)
           } else {
             groupingAttrs(index)
           }
-      }.asInstanceOf[NamedExpression]
+      }
+      aggregations.map { agg =>
+        replaceExprs(replaceGroupingFunc(agg, groupByExprs, gid)).asInstanceOf[NamedExpression]
+      }
```
[the remark this answers is not in the document]
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: +1.

**Review by jaceklaskowski — APPROVED** [review]

**Comment by wangyum** [comment]
Merged to master, branch-3.4 and branch-3.3.

**Pull request closed without merging**